### PR TITLE
[AMDGPU][SplitModule] Cleanup CallsExternal Handling

### PR DESCRIPTION
**Pull request by Pierre-vh**
- Don't treat inline ASM as indirect calls
- Remove call to alias testing, which was broken (only working by pure luck right now) and isn't needed anyway. GlobalOpt should take care of them for us.

### CODE REVIEW

**Comment by jmmartinez** [comment]
Would it make sense to replace the alias uses by the aliased global instead ? Such that all references (or at least the calls) inside a module point directly to the actual global and not to the alias.

File: llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp [hunk 1]
```diff
@@ -511,29 +523,68 @@ void SplitGraph::buildGraph(CallGraph &CG) {
...
-    if (HasIndirectCall) {
-      LLVM_DEBUG(dbgs() << "indirect call found in " << Fn.getName() << "\n");
-      FnsWithIndirectCalls.push_back(&Fn);
+    if (CallsExternal) {
+      LLVM_DEBUG(dbgs() << "  [!] callgraph is incomplete for " << Fn.getName()
+                        << " - analyzing function\n");
+
+      bool HasIndirectCall = false;
+      for (const auto &BB : Fn) {
+        for (const auto &Inst : BB) {
```
Comment (jmmartinez) [inline llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp:545]: nitpick: maybe use the inst-iterator

```suggestion
      for (const auto &Inst : instructions(&Fn)) {
```

File: llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp [hunk 2]
```diff
@@ -103,6 +103,15 @@ static cl::opt<bool> NoExternalizeGlobals(
...
              "may cause globals to be duplicated which increases binary size"));
 
+static cl::opt<bool> NoExternalizeOnAddrTaken(
+    "amdgpu-module-splitting-no-externalize-address-taken", cl::Hidden,
+    cl::desc(
+        "disables externalization of functions whose addresses are taken"));
+
+static cl::opt<bool> InlineAsmIsIndirectCall(
+    "amdgpu-module-splitting-inline-asm-is-indirect-call", cl::Hidden,
+    cl::desc("consider inline assembly as na indirect call"));
```
Comment (shiltian) [inline llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp:113]:
```suggestion
    cl::desc("consider inline assembly as an indirect call"));
```

Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp:113]: Nobody will ever use these flags, I would prefer to not add them

**Review by arsenm — COMMENTED** [review]
Other optimizations already redirect calls to aliases to direct calls, so I don't think this pass should have to put in the effort to also do the same. I think GlobalOpt?

File: llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp (hunk 1 above)
Comment (Pierre-vh) [reply llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp:545]: This doesn't seem to exist

Comment (jmmartinez) [reply llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp:545]: I meant this one, or did I miss something ?

https://github.com/llvm/llvm-project/blob/3d5e1ec6508c8425601d4cfaba4c8a8f18791e2b/llvm/include/llvm/IR/InstIterator.h#L133

Comment (Pierre-vh) [reply llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp]: No, but I missed it :D

**Comment by Pierre-vh** [comment]
> Other optimizations already redirect calls to aliases to direct calls, so I don't think this pass should have to put in the effort to also do the same. I think GlobalOpt?

Now that I remember, this is indeed an artificial case.
I'll check if just adding an opt run line also does the same thing, if it does, then I'll just eliminate that part of the alias handling.

**Comment by arsenm** [comment]
> Though I'm debating whether it's worth removing. It's just a few lines of code and it makes the pass more correct in case globalopt didn't run for some reason (not sure if it runs even at O0?), but it's also redundant code which isn't great

I'd rather just let global opt deal with it. It probably doesn't run at -O0

File: llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp [hunk 3]
```diff
@@ -486,6 +486,27 @@ void SplitGraph::Node::visitAllDependencies(
   }
 }
 
+/// Checks if \p I has MD_callees and if it does, parse it and put the function
+/// in \p Callees.
+///
+/// \returns true if there was metadata and it was parsed correctly. false if
+/// there was no MD or if it contained unknown entries.
+static bool handleCalleesMD(const Instruction &I,
+                            SmallVector<Function *> &Callees) {
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp:495]: I would expect the CallGraph analysis to parse this metadata for you. Also this would be a separate patch from "aliases and inline asm"

Comment (Pierre-vh) [reply llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp:495]: I thought of this patch more as a general improvement on the CallsExternal case in the CG.
Is it okay if I retitle the patch, or do you want me to split this into 2 or 3 patches ?

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp:495]: Split it

**Comment by Pierre-vh** [comment]
ping

File: llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp [hunk 4]
```diff
@@ -1326,13 +1362,23 @@ static void splitAMDGPUModule(
...
+  // externalizing like this. Maybe non-copyable should really mean "keep one
+  // visible copy, then internalize all other copies" for some functions?
+  if (!NoExternalizeOnAddrTaken) {
+    for (auto &Fn : M) {
+      // TODO: Should aliases count? Probably not but they're so rare I'm not
+      // sure it's worth fixing.
+      if (Fn.hasAddressTaken()) {
+        if (Fn.hasLocalLinkage()) {
+          LLVM_DEBUG(dbgs() << "[externalize] " << Fn.getName()
+                            << " because its address is taken\n");
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp:1378]: Should use printAsOperand to handle anonymous functions correctly

File: llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp [hunk 5]
```diff
@@ -1326,13 +1362,23 @@ static void splitAMDGPUModule(
...
+  // addresses are taken non-copyable. We should probably model this type of
+  // constraint in the graph and use it to guide splitting, instead of
+  // externalizing like this. Maybe non-copyable should really mean "keep one
+  // visible copy, then internalize all other copies" for some functions?
+  if (!NoExternalizeOnAddrTaken) {
+    for (auto &Fn : M) {
+      // TODO: Should aliases count? Probably not but they're so rare I'm not
+      // sure it's worth fixing.
+      if (Fn.hasAddressTaken()) {
+        if (Fn.hasLocalLinkage()) {
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp:1376]: Check hasLocalLinkage first

File: llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp [hunk 6]
```diff
@@ -103,6 +104,15 @@ static cl::opt<bool> NoExternalizeGlobals(
...
              "may cause globals to be duplicated which increases binary size"));
 
+static cl::opt<bool> NoExternalizeOnAddrTaken(
+    "amdgpu-module-splitting-no-externalize-address-taken", cl::Hidden,
+    cl::desc(
+        "disables externalization of functions whose addresses are taken"));
+
+static cl::opt<bool> InlineAsmIsIndirectCall(
+    "amdgpu-module-splitting-inline-asm-is-indirect-call", cl::Hidden,
+    cl::desc("consider inline assembly as an indirect call"));
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp:114]: I don't think this option is useful as defined, and it introduces an inconsistency with the attributor treatment. While probably incorrect, the attributor currently assumes inline asm doesn't contain a call of any kind. If we really want a cl::opt, it should be used everywhere 

We probably should fix that though; we even assume there's no call if the asm directly refers to a function as a pointer argument

File: llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp [hunk 7]
```diff
@@ -492,29 +505,52 @@ void SplitGraph::buildGraph(CallGraph &CG) {
...
+        CallsExternal = true;
     }
 
     // Keep track of this function if it contains an indirect call and/or if it
     // can be indirectly called.
-    if (HasIndirectCall) {
-      LLVM_DEBUG(dbgs() << "indirect call found in " << Fn.getName() << "\n");
-      FnsWithIndirectCalls.push_back(&Fn);
+    if (CallsExternal) {
+      LLVM_DEBUG(dbgs() << "  [!] callgraph is incomplete for " << Fn.getName()
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp:522]: printAsOperand

File: llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp [hunk 8]
```diff
@@ -492,29 +501,51 @@ void SplitGraph::buildGraph(CallGraph &CG) {
...
+            CB && !CB->getCalledFunction()) {
+          // inline assembly can be ignored, unless InlineAsmIsIndirectCall is
+          // true.
+          if (CB->isInlineAsm()) {
+            LLVM_DEBUG(dbgs() << "    found inline assembly\n");
+            continue;
+          }
+
+          // everything else is handled conservatively.
+          HasIndirectCall = true;
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp:535]: Can break at this point? Also can just turn this into a helper function to scan for indirect calls and use early return

**Review by arsenm — APPROVED** [review]
